src/OrchardCore/OrchardCore.Users.Core/Services/DefaultUserClaimsPrincipalProviderFactory.cs

```diff
@@ -0,0 +1,39 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.Extensions.Options;
+using OrchardCore.Security;
+
+namespace OrchardCore.Users.Services
+{
+    /// <summary>
+    /// Custom implementation of <see cref="IUserClaimsPrincipalFactory{TUser}"/> allowing adding claims by implementing the <see cref="IClaimsProvider"/>
+    /// </summary>
+    public class DefaultUserClaimsPrincipalProviderFactory : UserClaimsPrincipalFactory<IUser, IRole>
+    {
+        private readonly IEnumerable<IClaimsProvider> _claimsProviders;
+
+        public DefaultUserClaimsPrincipalProviderFactory(
+            UserManager<IUser> userManager,
+            RoleManager<IRole> roleManager,
+            IOptions<IdentityOptions> identityOptions,
+            IEnumerable<IClaimsProvider> claimsProviders) : base(userManager, roleManager, identityOptions)
+        {
+            _claimsProviders = claimsProviders ?? Enumerable.Empty<IClaimsProvider>();
```

Comment:
The DI Container will supply an empty enumerable automatically. So no need for this check.

Reply:
But what if a user constructed this class outside the DI? like for testing purposes?

Reply:
Then they have written a bad test and should write one using the correct signature. We've had this discussion before in OC, we don't need to do this for DI constructors.

I unresolved the other comments that I made, as they still need to be attended to